server/core_authenticate.go

```diff
@@ -275,6 +276,44 @@ func AuthenticateEmail(ctx context.Context, logger *zap.Logger, db *sql.DB, emai
 	return userID, username, true, nil
 }
 
+func AuthenticateUsername(ctx context.Context, logger *zap.Logger, db *sql.DB, username, password string) (string, error) {
+	// Look for an existing account.
+	query := "SELECT id, password, disable_time FROM users WHERE username = $1"
+	var dbUserID string
+	var dbPassword []byte
+	var dbDisableTime pq.NullTime
+	err := db.QueryRowContext(ctx, query, username).Scan(&dbUserID, &dbPassword, &dbDisableTime)
+	if err != nil {
+		if err == sql.ErrNoRows {
+			// Account not found and creation is never allowed for this type.
+			return "", status.Error(codes.NotFound, "User account not found.")
+		} else {
+			logger.Error("Error looking up user by username.", zap.Error(err), zap.String("username", username))
+			return "", status.Error(codes.Internal, "Error finding user account.")
+		}
+	}
+
+	// Check if it's disabled.
+	if dbDisableTime.Valid && dbDisableTime.Time.Unix() != 0 {
+		logger.Info("User account is disabled.", zap.String("username", username))
+		return "", status.Error(codes.Unauthenticated, "Error finding or creating user account.")
```

Comment:
Maybe better to return more clear error description, the same as logged?

Reply:
This is intentionally vague. The developer gets a descriptive log message but under the current API contract clients are not told when their account is disabled.